### PR TITLE
Added list dataset

### PR DESCRIPTION
**Pull request by churnikov**
For now, we don't push you to follow any predefined schema of issue, but ensure you've already read our contribution guide: https://open-metric-learning.readthedocs.io/en/latest/from_readme/contributing.html.

### CODE REVIEW

File: oml/datasets/base.py [hunk 1]
```diff
@@ -173,6 +175,84 @@ def bboxes_keys(self) -> Tuple[str, ...]:
...
+            if self.bboxes is not None:
+                for bbox in self.bboxes[i]:
+                    yield self.apply_transforms(img, bbox)
+            else:
+                img = self.f_imread(img_bytes)
+                im_h, im_w = img.shape[:2] if isinstance(img, np.ndarray) else img.size[::-1]
+                bbox = (0, 0, im_w, im_h)
+                yield self.apply_transforms(img, bbox)
+
+    def __len__(self) -> int:
```
Comment (DaloroAT) [inline oml/datasets/base.py:243]: suppose it is `return len(self.filenames_list)`. Bboxes are not related to len of dataset

File: tests/test_oml/test_datasets/test_list_dataest.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+from oml.const import MOCK_DATASET_PATH
+from oml.datasets.base import ListDataset
+
+
+def test_dataset_len() -> None:
+    images = list((MOCK_DATASET_PATH / "images").iterdir())
+    dataset = ListDataset(images)
+
+    assert len(dataset) == len(images)
```
Comment (DaloroAT) [inline tests/test_oml/test_datasets/test_list_dataest.py:11]: add `assert len(images) > 0`. If you have no images this test will be passed

File: tests/test_oml/test_datasets/test_list_dataest.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+
+def test_dataset_iter() -> None:
```
Comment (DaloroAT) [inline tests/test_oml/test_datasets/test_list_dataest.py:14]: please, iterate over dataset with `Dataloader`. And check the shape of the batch is `ndim==4` and that all dims are not zero.

File: tests/test_oml/test_datasets/test_list_dataest.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import torch
```
Comment (DaloroAT) [inline tests/test_oml/test_datasets/test_list_dataest.py:1]: There are no tests for a case with bounding boxes. Please add a test according to the new recommendation of `test_dataset_iter` test.

You can find csv with bounding boxes in `MOCK_DATASET_PATH / "df_with_bboxes.csv". Or you can create random bboxes.

You also can parameterize `test_dataset_iter` test with bounding boxes. It's up to you.

**Review by DaloroAT — CHANGES_REQUESTED** [review]

File: oml/datasets/base.py [hunk 5]
```diff
@@ -173,6 +175,84 @@ def bboxes_keys(self) -> Tuple[str, ...]:
...
+            else:
+                # torchvision.transforms
+                image_tensor = self.transform(img)
+        image_tensor = torch.tensor(img)
+
+        return image_tensor
+
+    def __iter__(self) -> Iterator[torch.Tensor]:
+        for i, im_path in enumerate(self.filenames_list):
+            img_bytes = self.read_bytes_image_cached(im_path)
```
[the remark this answers is not in the document]
Comment (churnikov) [reply oml/datasets/base.py:231]: But I thought that you can implement either `__getitem` or `__iter__`. And if we keep this: https://github.com/OML-Team/open-metric-learning/pull/210/files#r1009367203 `__iter__` in that case is easier to work with

File: oml/datasets/base.py [hunk 6]
```diff
@@ -173,6 +175,84 @@ def bboxes_keys(self) -> Tuple[str, ...]:
...
+        if isinstance(img, Image.Image):
+            return np.array(img.crop(bbox))
+        else:
+            x1, y1, x2, y2 = bbox
+            return img[y1:y2, x1:x2, :]
+
+    def apply_transforms(self, img: Union[Image.Image], bbox: Tuple[int, int, int, int]) -> torch.Tensor:
+        x1, y2, x2, y2 = bbox
+        img = self.crop(img, bbox)  # todo: since albu may handle bboxes we should move it to augs
+        if self.transform is not None:
```
[the remark this answers is not in the document]
Comment (DaloroAT) [reply oml/datasets/base.py:219]: Sorry, I should have added a [link](https://github.com/OML-Team/open-metric-learning/pull/210/files#r1009389322) to idea

File: oml/datasets/base.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (DaloroAT) [reply oml/datasets/base.py:231]: Oh, I got your idea. You wanted to make it possible to use multiple boxes for each picture. Let's discuss this in a separate comment

**Comment by DaloroAT** [comment]
@churnikov 

We are discussing the possibility to use several bounding boxes for each image.
How we can implement this idea:
1) Filenames are unique, for each image we send list of boxes to ListDataset https://github.com/OML-Team/open-metric-learning/pull/210#discussion_r1010309582
2) Prepare list of non-unique filenames outside of ListDataset and flat nested bounding boxes

For me, 2 approach is better:
1) We can implement `__getitem__` for sample `(image, bbox)`
2) We don't need to implement nested calculations in datasets

What do you think @churnikov ?

File: oml/datasets/base.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (DaloroAT) [reply oml/datasets/base.py:231]: https://github.com/OML-Team/open-metric-learning/pull/210#issuecomment-1298615973

File: oml/datasets/base.py [hunk 7]
```diff
@@ -173,6 +175,84 @@ def bboxes_keys(self) -> Tuple[str, ...]:
             return tuple()
 
 
+class ListDataset(Dataset):
+    """Iterate over list of images"""
+
+    def __init__(
+        self,
+        filenames_list: Sequence[Path],
+        bboxes: Optional[Sequence[Sequence[Tuple[int, int, int, int]]]] = None,
```
[the remark this answers is not in the document]
Comment (DaloroAT) [reply oml/datasets/base.py:184]: https://github.com/OML-Team/open-metric-learning/pull/210#issuecomment-1298615973

**Comment by churnikov** [comment]
I can implement the second one, but I think that it could be a bit unintuitive. On the other hand, the first approach has a drawback right now of having potentially a lot of empty nested lists (when there are no boxes for image). But it could be solved with replacement of empty lists to `None`. And I agree that `__getitem__` is better

But if it's just me, and you think that the second approach is intuitive, then I'll do it of course

**Comment by AlekseySh** [comment]
@churnikov the rest of the repo assumes that 1 line in the `df.csv` is one box (object), so the second approach fits OML better

**Comment by churnikov** [comment]
Okay, I'll redo it tomorrow :)

**Comment by churnikov** [comment]
Hi!

I've applied requested changes :)

File: oml/datasets/list_.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+                    x1, y1, x2, y2 = box
+                    if x2 <= x1 or y2 <= y1:
+                        raise InvalidBBoxesException(f"Bbox has invalid dimensions for image {files[idx]}")
+
+    @staticmethod
+    def _read_bytes_image(path: Union[Path, str]) -> bytes:
+        with open(str(path), "rb") as fin:
+            return fin.read()
+
+    def crop(self, img: Union[Image.Image, np.ndarray], bbox: TBBox) -> np.ndarray:
```
Comment (AlekseySh) [inline oml/datasets/list_.py:69]: there is no need to create a dedicated function for cropping, in this case, code will be more readable if you just repeat the logic from here https://github.com/OML-Team/open-metric-learning/blob/main/oml/datasets/base.py#L136

File: oml/datasets/list_.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+
+    def __getitem__(self, idx: int) -> torch.Tensor:
+        im_path = self.filenames_list[idx]
+        img_bytes = self.read_bytes_image_cached(im_path)
+        img = self.f_imread(img_bytes)
+        bbox = self.bboxes.get(idx)
+
+        if bbox is None:
+            img_arr = np.array(img)
+            print(img_arr.shape)
```
Comment (AlekseySh) [inline oml/datasets/list_.py:95]: rm print

File: oml/datasets/list_.py [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+        return image_tensor
+
+    def __getitem__(self, idx: int) -> torch.Tensor:
+        im_path = self.filenames_list[idx]
+        img_bytes = self.read_bytes_image_cached(im_path)
+        img = self.f_imread(img_bytes)
+        bbox = self.bboxes.get(idx)
+
+        if bbox is None:
+            img_arr = np.array(img)
```
Comment (AlekseySh) [inline oml/datasets/list_.py:94]: let's avoid changing the type of the image without the need
see the example here https://github.com/OML-Team/open-metric-learning/blob/main/oml/datasets/base.py#L127

File: oml/datasets/list_.py [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+            return fin.read()
+
+    def crop(self, img: Union[Image.Image, np.ndarray], bbox: TBBox) -> np.ndarray:
+        if isinstance(img, Image.Image):
+            return np.array(img.crop(bbox))
+        else:
+            x1, y1, x2, y2 = bbox
+            return img[y1:y2, x1:x2, :]
+
+    def apply_transforms(self, img: Union[Image.Image], bbox: TBBox) -> torch.Tensor:
```
Comment (AlekseySh) [inline oml/datasets/list_.py:76]: I guess there is not need in dedicated function here
mv its content to the `get_item` please

File: tests/test_oml/test_datasets/test_list_dataest.py [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+from oml.const import MOCK_DATASET_PATH
+from oml.datasets.list_ import ListDataset
+
+
+@pytest.fixture
+def images() -> Iterator[List[Path]]:
+    yield list((MOCK_DATASET_PATH / "images").iterdir())
+
+
+def get_images_and_boxes() -> List[Tuple[Path, int, int, int, int]]:
```
Comment (AlekseySh) [inline tests/test_oml/test_datasets/test_list_dataest.py:18]: let's use `pd.read_csv` and then call `df['x1']..` instead of unpacking based on positions, which is dangerous and less readable

File: tests/test_oml/test_datasets/test_list_dataest.py [hunk 13]
```diff
@@ -0,0 +1,57 @@
+import csv
```
Comment (AlekseySh) [inline tests/test_oml/test_datasets/test_list_dataest.py:1]: as for tests in general, I believe we need one more test, when part of the images have boxes, and part of them don't

**Review by AlekseySh — CHANGES_REQUESTED** [review]

File: oml/datasets/list_.py [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+TBBox = Union[List[int], Tuple[int, int, int, int]]
+
+
+class ListDataset(Dataset):
+    """Iterate over list of images"""
+
+    def __init__(
+        self,
+        filenames_list: Sequence[Path],
+        bboxes: Optional[Dict[int, TBBox]] = None,
```
[the remark this answers is not in the document]
Comment (churnikov) [reply oml/datasets/list_.py:25]: Okay, I was just considering case where there are more boxes missing than existing. But I guess the opposite is probably expected.

File: oml/datasets/list_.py [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+    ):
+        """
+        Args:
+            filenames_list: list of paths to images
+            boxes: Sequences of bounding boxes. Should be either ``None`` or
+                Sequence of of bboxes. If the image has multiple boxes, pass
+                multiple image paths to ``filenames_list`` and for each
+                filename provide one of its boxes. If you want to get
+                embeddings for the whole image, set bbox to ``None`` for
+                specific file.
```
[the remark this answers is not in the document]
Comment (churnikov) [reply oml/datasets/list_.py:38]: I wrote this as I don't think that it's obvious, but maybe that's just me. If you insist, I'll remove it of course :)

File: oml/datasets/list_.py [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+import torch
+from PIL import Image
+from torch.utils.data import Dataset
+
+from oml.exceptions import InvalidBBoxesException
+from oml.transforms.images.torchvision.transforms import get_normalisation_torch
+from oml.transforms.images.utils import TTransforms
+from oml.utils.images.images import TImReader, imread_cv2
+
+TBBox = Union[List[int], Tuple[int, int, int, int]]
```
[the remark this answers is not in the document]
Comment (churnikov) [reply oml/datasets/list_.py:16]: On one hand, I agree, on the other, It was @DaloroAT idea
https://github.com/OML-Team/open-metric-learning/pull/210#discussion_r1009367203

And I actually agree with it, considering how you define bboxes in this comment:
https://github.com/OML-Team/open-metric-learning/pull/210/files#r1014853658

File: oml/datasets/list_.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (AlekseySh) [reply oml/datasets/list_.py:38]: I don't insist, we can keep it

File: oml/datasets/list_.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (AlekseySh) [reply oml/datasets/list_.py:16]: In that comment @DaloroAT most likely just tried to obtain the correct typing from the existing code rather than suggesting the idea of the type. I cannot ask him in person, because he will be on vacation the next days. So, let's keep the simple one `Tuple[int, int, int, int]`

File: oml/datasets/list_.py (hunk 8 above)
Comment (churnikov) [reply oml/datasets/list_.py:69]: I actually have a question to that code. Torchvision  can accept PIL and Tensor images, so I think that It's wrong to assume that usage of torch vision leads to usage of PIL.

https://pytorch.org/vision/stable/transforms.html#transforms-on-torch-tensor-only

> Most transformations accept both [PIL](https://pillow.readthedocs.io/) images and tensor images, although some transformations are [PIL-only](https://pytorch.org/vision/stable/transforms.html#transforms-pil-only) and some are [tensor-only](https://pytorch.org/vision/stable/transforms.html#transforms-tensor-only).

Comment (AlekseySh) [reply oml/datasets/list_.py:73]: @churnikov yes, but it does not work with NumPy.
So, our separation is the following: `cv2 reader -> numpy -> albu` or `pil_reader -> pil -> torchtransforms`

But it's not really related to the current comment. The idea was just to simplify the code and put it to `getitem`

Comment (AlekseySh) [reply oml/datasets/list_.py:73]: @churnikov yes, but it does not work with NumPy.
So, our separation is the following: `cv2 reader -> numpy -> albu` or `pil_reader -> pil -> torchtransforms`

But it's not really related to the current comment. The idea was just to simplify the code and put it to `getitem`

**Comment by AlekseySh** [comment]
@churnikov thx for the updates, let us know in comments when your PR is ready to check

**Comment by churnikov** [comment]
Hi!

I think I did everything suggested 😊

File: oml/datasets/list_.py [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+            bbox = self.bboxes[idx]
+        else:
+            bbox = None
+
+        if bbox is None:
+            im_h, im_w = img.shape[:2] if isinstance(img, np.ndarray) else img.size[::-1]
+            bbox = (0, 0, im_w, im_h)
+
+        if isinstance(img, Image.Image):
+            img = np.array(img.crop(bbox))
```
Comment (AlekseySh) [inline oml/datasets/list_.py:87]: Don't convert the image to NumPy. The idea is that reader has to be aligned with the given transformation. 
(See BaseDataset for an example)

File: tests/test_oml/test_datasets/test_list_dataest.py [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+def images() -> Iterator[List[Path]]:
+    yield list((MOCK_DATASET_PATH / "images").iterdir())
+
+
+def get_images_and_boxes() -> List[Tuple[Path, int, int, int, int]]:
+    df = pd.read_csv(MOCK_DATASET_PATH / "df_with_bboxes.csv")
+    sub_df = df[["path", "x_1", "y_1", "x_2", "y_2"]]
+    sub_df["path"] = sub_df.path.apply(lambda p: MOCK_DATASET_PATH / p)
+    lines: List[Tuple[Path, int, int, int, int]] = [tuple(line[1]) for line in sub_df.iterrows()]  # type: ignore
+    return lines
```
Comment (AlekseySh) [inline tests/test_oml/test_datasets/test_list_dataest.py:23]: To make it more real let's return the structures that we pass to the dataset: list of paths and boxes (`Sequence[Optional[TBBox]]`)

Comment (AlekseySh) [reply tests/test_oml/test_datasets/test_list_dataest.py:23]: see the comment below ->

File: tests/test_oml/test_datasets/test_list_dataest.py [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+
+def test_dataloader_iter(images: List[Path]) -> None:
+    dataloader = DataLoader(ListDataset(images))
+
+    for im in dataloader:
+        assert im.ndim == 4
+
+
+@pytest.mark.parametrize("im_path,x1,y1,x2,y2", get_images_and_boxes())
+def test_mock_dataset_iter(im_path: Path, x1: int, x2: int, y1: int, y2: int) -> None:
```
Comment (AlekseySh) [inline tests/test_oml/test_datasets/test_list_dataest.py:48]: -> see comment above
and let's pass the structure from the `get_images_and_boxes` directly to the dataset, not line by line (it also means that you don't need parameterize)

because now we create a dataset based on 1 image and 1 bbox, which is not realistic

File: tests/test_oml/test_datasets/test_list_dataest.py [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+    dataloader = DataLoader(ListDataset([im_path], [(x1, y1, x2, y2)]))
+
+    image = next(iter(dataloader))
+    assert image.size() == (1, 3, x2 - x1, y2 - y1)
+
+
+def test_mock_dataset_iter_with_nones() -> None:
```
Comment (AlekseySh) [inline tests/test_oml/test_datasets/test_list_dataest.py:55]: rename to `test_list_dataset_with_some_nones` (same logic)

File: tests/test_oml/test_datasets/test_list_dataest.py (hunk 19 above)
Comment (AlekseySh) [inline tests/test_oml/test_datasets/test_list_dataest.py:48]: rename to `test_list_dataset_iter` (because we don't test mock dataset here, we test a python class)

**Review by AlekseySh — CHANGES_REQUESTED** [review]
Good job, I guess, we are almost finished.

Please, don't forget to add the new class into docs (see `docs/source/contents/datasets`)

File: oml/datasets/list_.py (hunk 17 above)
Comment (churnikov) [reply oml/datasets/list_.py:87]: Sorry, I thought I changed it

**Comment by churnikov** [comment]
To the top 🚀 

Going to get to top 3 most discussed PR :))

**Comment by AlekseySh** [comment]
@churnikov Some of the PRs also had a long history of discussion, but they were conducted offline :)

**Comment by AlekseySh** [comment]
@churnikov Everything is fine now, but you missed my message:
>>> Please, don't forget to add the new class into docs (see docs/source/contents/datasets)

**Review by AlekseySh — APPROVED** [review]

**Comment by AlekseySh** [comment]
@churnikov congratulations with the 1st contribution to OML!
